erts/mac/man/man3/CosTransactions_Resource.3
.TH CosTransactions_Resource 3 "cosTransactions 1.3" "Ericsson AB" "Erlang Module Definition"
.SH NAME
CosTransactions_Resource \- This module implements the OMG CosTransactions::Resource interface.
.SH DESCRIPTION
.LP
To get access to the record definitions for the structures use: 
.br
\fI-include_lib("cosTransactions/include/CosTransactions\&.hrl")\&.\fR\&
.SH EXPORTS
.LP
.B
commit(Resource) -> Return
.br
.RS
.LP
Types:

.RS 3
Resource = #objref
.br
Return = ok | {\&'EXCEPTION\&', E}
.br
E = #\&'CosTransactions_NotPrepared\&'{} | #\&'CosTransactions_HeuristicRollback\&'{} | #\&'CosTransactions_HeuristicMixed\&'{} | #\&'CosTransactions_HeuristicHazard\&'{}
.br
.RE
.RE
.RS
.LP
This operation instructs the Resource to commit all changes made as a part of the transaction\&.
.LP
The Resource can raise:
.RS 2
.TP 2
*
Heuristic Exception - if a Heuristic decision is made which differ from the true outcome of the transaction\&. The Resource must remember the Heuristic outcome until the \fIforget\fR\& operation is performed\&.
.LP
.RE

.RE
.LP
.B
commit_one_phase(Resource) -> Return
.br
.RS
.LP
Types:

.RS 3
Resource = #objref
.br
Return = ok | {\&'EXCEPTION\&', E}
.br
E = #\&'CosTransactions_HeuristicHazard\&'{} | #\&'CosTransactions_TransactionRolledBack\&'{}
.br
.RE
.RE
.RS
.LP
If possible, the Resource should commit all changes made as part of the transaction, else it should raise the TRANSACTION_ROLLEDBACK exception\&. This operation can only be used if the Resource is the only child of its parent\&.
.RE
.LP
.B
forget(Resource) -> Return
.br
.RS
.LP
Types:

.RS 3
Resource = #objref
.br
Return = ok
.br
.RE
.RE
.RS
.LP
This operation informs the Resource that it is safe to forget any Heuristic decisions and the knowledge of the transaction\&.
.RE
.LP
.B
prepare(Resource) -> Return
.br
.RS
.LP
Types:

.RS 3
Resource = #objref
.br
Return = Vote | {\&'EXCEPTION\&', E}
.br
Vote = \&'VoteReadOnly\&' | \&'VoteCommit\&' | \&'VoteRollback\&'
.br
E = #\&'CosTransactions_HeuristicMixed\&'{} | #\&'CosTransactions_HeuristicHazard\&'{}
.br
.RE
.RE
.RS
.LP
This operation is invoked on the Resource to begin the two-phase commit protocol\&.
.LP
The Resource can reply:
.RS 2
.TP 2
*
\&'VoteReadOnly\&' - if no persistent data has been modified by the transaction\&. The Resource can forget all knowledge of the transaction\&. 
.LP
.TP 2
*
\&'VoteCommit\&' - if the Resource has been prepared and is able to write all the data needed to commit the transaction to stable storage\&.
.LP
.TP 2
*
\&'VoteRollback\&' - under any circumstances but must do so if none of the alternatives above are applicable\&.
.LP
.TP 2
*
Heuristic Exception - if a Heuristic decision is made which differ from the true outcome of the transaction\&. The Resource must remember the Heuristic outcome until the \fIforget\fR\& operation is performed\&.
.LP
.RE

.RE
.LP
.B
rollback(Resource) -> Return
.br
.RS
.LP
Types:

.RS 3
Resource = #objref
.br
Return = ok | {\&'EXCEPTION\&', E}
.br
E = #\&'CosTransactions_HeuristicCommit\&'{} | #\&'CosTransactions_HeuristicMixed\&'{} | #\&'CosTransactions_HeuristicHazard\&'{}
.br
.RE
.RE
.RS
.LP
This operation instructs the Resource to rollback all changes made as a part of the transaction\&.
.LP
The Resource can raise:
.RS 2
.TP 2
*
Heuristic Exception - if a Heuristic decision is made which differ from the true outcome of the transaction\&. The Resource must remember the Heuristic outcome until the \fIforget\fR\& operation is performed\&.
.LP
.RE

.RE

erts/mac/lib/cosTransactions-1.3/src/CosTransactions_Resource.erl
%%  coding: latin-1
%%------------------------------------------------------------
%%
%% Implementation stub file
%% 
%% Target: CosTransactions_Resource
%% Source: /private/tmp/erlang20150820-34163-1k9di6g/otp-OTP-18.0.3/lib/cosTransactions/src/CosTransactions.idl
%% IC vsn: 4.4
%% 
%% This file is automatically generated. DO NOT EDIT IT.
%%
%%------------------------------------------------------------

-module('CosTransactions_Resource').
-ic_compiled("4_4").


%% Interface functions
-export([prepare/1, prepare/2, rollback/1]).
-export([rollback/2, commit/1, commit/2]).
-export([commit_one_phase/1, commit_one_phase/2, forget/1]).
-export([forget/2]).

%% Type identification function
-export([typeID/0]).

%% Used to start server
-export([oe_create/0, oe_create_link/0, oe_create/1]).
-export([oe_create_link/1, oe_create/2, oe_create_link/2]).

%% TypeCode Functions and inheritance
-export([oe_tc/1, oe_is_a/1, oe_get_interface/0]).

%% gen server export stuff
-behaviour(gen_server).
-export([init/1, terminate/2, handle_call/3]).
-export([handle_cast/2, handle_info/2, code_change/3]).

-include_lib("orber/include/corba.hrl").


%%------------------------------------------------------------
%%
%% Object interface functions.
%%
%%------------------------------------------------------------



%%%% Operation: prepare
%% 
%%   Returns: RetVal
%%   Raises:  CosTransactions::HeuristicMixed, CosTransactions::HeuristicHazard
%%
prepare(OE_THIS) ->
    corba:call(OE_THIS, prepare, [], ?MODULE).

prepare(OE_THIS, OE_Options) ->
    corba:call(OE_THIS, prepare, [], ?MODULE, OE_Options).

%%%% Operation: rollback
%% 
%%   Returns: RetVal
%%   Raises:  CosTransactions::HeuristicCommit, CosTransactions::HeuristicMixed, CosTransactions::HeuristicHazard
%%
rollback(OE_THIS) ->
    corba:call(OE_THIS, rollback, [], ?MODULE).

rollback(OE_THIS, OE_Options) ->
    corba:call(OE_THIS, rollback, [], ?MODULE, OE_Options).

%%%% Operation: commit
%% 
%%   Returns: RetVal
%%   Raises:  CosTransactions::NotPrepared, CosTransactions::HeuristicRollback, CosTransactions::HeuristicMixed, CosTransactions::HeuristicHazard
%%
commit(OE_THIS) ->
    corba:call(OE_THIS, commit, [], ?MODULE).

commit(OE_THIS, OE_Options) ->
    corba:call(OE_THIS, commit, [], ?MODULE, OE_Options).

%%%% Operation: commit_one_phase
%% 
%%   Returns: RetVal
%%   Raises:  CosTransactions::HeuristicHazard, CosTransactions::HeuristicRollback, CosTransactions::HeuristicMixed
%%
commit_one_phase(OE_THIS) ->
    corba:call(OE_THIS, commit_one_phase, [], ?MODULE).

commit_one_phase(OE_THIS, OE_Options) ->
    corba:call(OE_THIS, commit_one_phase, [], ?MODULE, OE_Options).

%%%% Operation: forget
%% 
%%   Returns: RetVal
%%
forget(OE_THIS) ->
    corba:call(OE_THIS, forget, [], ?MODULE).

forget(OE_THIS, OE_Options) ->
    corba:call(OE_THIS, forget, [], ?MODULE, OE_Options).

%%------------------------------------------------------------
%%
%% Inherited Interfaces
%%
%%------------------------------------------------------------
oe_is_a("IDL:omg.org/CosTransactions/Resource:1.0") -> true;
oe_is_a(_) -> false.

%%------------------------------------------------------------
%%
%% Interface TypeCode
%%
%%------------------------------------------------------------
oe_tc(prepare) -> 
	{{tk_enum,"IDL:omg.org/CosTransactions/Vote:1.0","Vote",
                  ["VoteCommit","VoteRollback","VoteReadOnly"]},
         [],[]};
oe_tc(rollback) -> 
	{tk_void,[],[]};
oe_tc(commit) -> 
	{tk_void,[],[]};
oe_tc(commit_one_phase) -> 
	{tk_void,[],[]};
oe_tc(forget) -> 
	{tk_void,[],[]};
oe_tc(_) -> undefined.

oe_get_interface() -> 
	[{"forget", oe_tc(forget)},
	{"commit_one_phase", oe_tc(commit_one_phase)},
	{"commit", oe_tc(commit)},
	{"rollback", oe_tc(rollback)},
	{"prepare", oe_tc(prepare)}].




%%------------------------------------------------------------
%%
%% Object server implementation.
%%
%%------------------------------------------------------------


%%------------------------------------------------------------
%%
%% Function for fetching the interface type ID.
%%
%%------------------------------------------------------------

typeID() ->
    "IDL:omg.org/CosTransactions/Resource:1.0".


%%------------------------------------------------------------
%%
%% Object creation functions.
%%
%%------------------------------------------------------------

oe_create() ->
    corba:create(?MODULE, "IDL:omg.org/CosTransactions/Resource:1.0").

oe_create_link() ->
    corba:create_link(?MODULE, "IDL:omg.org/CosTransactions/Resource:1.0").

oe_create(Env) ->
    corba:create(?MODULE, "IDL:omg.org/CosTransactions/Resource:1.0", Env).

oe_create_link(Env) ->
    corba:create_link(?MODULE, "IDL:omg.org/CosTransactions/Resource:1.0", Env).

oe_create(Env, RegName) ->
    corba:create(?MODULE, "IDL:omg.org/CosTransactions/Resource:1.0", Env, RegName).

oe_create_link(Env, RegName) ->
    corba:create_link(?MODULE, "IDL:omg.org/CosTransactions/Resource:1.0", Env, RegName).

%%------------------------------------------------------------
%%
%% Init & terminate functions.
%%
%%------------------------------------------------------------

init(Env) ->
%% Call to implementation init
    corba:handle_init('ETraP_Server_impl', Env).

terminate(Reason, State) ->
    corba:handle_terminate('ETraP_Server_impl', Reason, State).


%%%% Operation: prepare
%% 
%%   Returns: RetVal
%%   Raises:  CosTransactions::HeuristicMixed, CosTransactions::HeuristicHazard
%%
handle_call({OE_THIS, OE_Context, prepare, []}, _, OE_State) ->
  corba:handle_call('ETraP_Server_impl', prepare, [], OE_State, OE_Context, OE_THIS, false);

%%%% Operation: rollback
%% 
%%   Returns: RetVal
%%   Raises:  CosTransactions::HeuristicCommit, CosTransactions::HeuristicMixed, CosTransactions::HeuristicHazard
%%
handle_call({OE_THIS, OE_Context, rollback, []}, _, OE_State) ->
  corba:handle_call('ETraP_Server_impl', rollback, [], OE_State, OE_Context, OE_THIS, false);

%%%% Operation: commit
%% 
%%   Returns: RetVal
%%   Raises:  CosTransactions::NotPrepared, CosTransactions::HeuristicRollback, CosTransactions::HeuristicMixed, CosTransactions::HeuristicHazard
%%
handle_call({OE_THIS, OE_Context, commit, []}, _, OE_State) ->
  corba:handle_call('ETraP_Server_impl', commit, [], OE_State, OE_Context, OE_THIS, false);

%%%% Operation: commit_one_phase
%% 
%%   Returns: RetVal
%%   Raises:  CosTransactions::HeuristicHazard, CosTransactions::HeuristicRollback, CosTransactions::HeuristicMixed
%%
handle_call({OE_THIS, OE_Context, commit_one_phase, []}, _, OE_State) ->
  corba:handle_call('ETraP_Server_impl', commit_one_phase, [], OE_State, OE_Context, OE_THIS, false);

%%%% Operation: forget
%% 
%%   Returns: RetVal
%%
handle_call({OE_THIS, OE_Context, forget, []}, _, OE_State) ->
  corba:handle_call('ETraP_Server_impl', forget, [], OE_State, OE_Context, OE_THIS, false);



%%%% Standard gen_server call handle
%%
handle_call(stop, _, State) ->
    {stop, normal, ok, State};

handle_call(_, _, State) ->
    {reply, catch corba:raise(#'BAD_OPERATION'{minor=1163001857, completion_status='COMPLETED_NO'}), State}.


%%%% Standard gen_server cast handle
%%
handle_cast(stop, State) ->
    {stop, normal, State};

handle_cast(_, State) ->
    {noreply, State}.


%%%% Standard gen_server handles
%%
handle_info(_, State) ->
    {noreply, State}.


code_change(OldVsn, State, Extra) ->
    corba:handle_code_change('ETraP_Server_impl', OldVsn, State, Extra).

